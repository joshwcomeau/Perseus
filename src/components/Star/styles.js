import { StyleSheet } from '../../helpers/global-aphrodite';

export const fadeInAnimation = {
  from: { opacity: 0 },
  to: { opacity: 1 }
};

export const twinkleAnimations = {
  quickTwinkle: {
    '0%': { opacity: 0.8 },
    '90%': { opacity: 0.8 },
    '92%': { opacity: 0.1 },
    '100%': { opacity: 0.8 },
  },
  slowFade: {
    '0%': { opacity: 0.75 },
    '25%': { opacity: 0 },
    '40%': { opacity: 0 },
    '100%': { opacity: 0.75 },
  },
  pulse: {
    '0%': { opacity: 0.5},
    '20%': { opacity: 0.2 },
    '30%': { opacity: 1 },
    '50%': { opacity: 0.3 },
    '65%': { opacity: 0.5 },
    '100%': { opacity: 0.5 },
  },
  dormant: {
    '0%': { opacity: 0.4 },
    '65%': { opacity: 0.4 },
    '80%': { opacity: 0.1 },
    '100%': { opacity: 0.4 },
  },
  blinkOut: {
    '0%': { opacity: 0.5 },
    '50%': { opacity: 0.5 },
    '55%': { opacity: 0 },
    '100%': { opacity: 0.5 },
  },
  sideShift: {
    '0%': { opacity: 0.5, transform: 'translateX(-6px)' },
    '50%': { opacity: 0.5, transform: 'translateX(0)' },
    '100%': { opacity: 0.5, transform: 'translateX(-6px)' },
  },
  bright: {
    '0%': { opacity: 0.2},
    '10%': { opacity: 1 },
    '25%': { opacity: 0.2 },
    '100%': { opacity: 0.2 },
  },
  grower: {
    '0%': { opacity: 0.5, transform: 'scale(1)' },
    '50%': { opacity: 0.75, transform: 'scale(2)' },
    '100%': { opacity: 0.5, transform: 'scale(1)' },
  },
};

const animationStyles = Object.keys(twinkleAnimations).reduce((styles, name) => ({
  [name]: { animationName: twinkleAnimations[name]},
  ...styles
}));


export default StyleSheet.create({
  starContainer: {
    position: 'absolute',
  },
  star: {
    animationIterationCount: 'infinite',
    animationFillMode: 'both',
    animationTimingFunction: 'linear',
  },
  fadeIn: {
    animationName: fadeInAnimation,
    animationIterationCount: 1,
    animationFillMode: 'both',
  },
  ...animationStyles,
});
