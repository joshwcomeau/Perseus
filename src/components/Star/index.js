// eslint-disable-next-line no-unused-vars
import React, { Component, PropTypes } from 'react';
import { css } from '../../helpers/global-aphrodite';
import random from 'lodash/random';
import sample from 'lodash/sample';

import styles, { fadeInAnimation, twinkleAnimations } from './styles';


const Star = ({ id, size, top, left, blurAmount, haloWidth, animationStyle }) => {
  const padding = size * 2; // Generous padding so that the star's blur isn't cut off
  const twinkleAnimationName = sample(Object.keys(twinkleAnimations));

  return (
    <div
      className={css(styles.starContainer, styles.fadeIn)}
      style={{
        top,
        left,
        animationDuration: `${random(0, 4000)}ms`,
        animationDelay: `${random(0, 1000)}ms`,
      }}
    >
      <svg
        className={css(styles.star, styles[twinkleAnimationName])}
        style={{
          animationDuration: `${random(5, 20, 0.1)}s`,
          animationDelay: `${random(0, 3000)}ms`,
        }}
        width={size + padding}
        height={size + padding}
      >
        <filter id={`blur-${id}`} x="-100%" y="-100%" width="1000%" height="1000%">
          <feGaussianBlur in="SourceGraphic" stdDeviation={blurAmount} />
        </filter>

        <circle
          cx={size / 2 + padding / 2}
          cy={size / 2 + padding / 2}
          r={size / 2}
          fill="white"
          stroke="#281f60"
          strokeWidth={haloWidth}
          filter={`url(#blur-${id})`}
        />
      </svg>
    </div>
  );
};

Star.propTypes = {

};

Star.defaultProps = {

};

export default Star;
