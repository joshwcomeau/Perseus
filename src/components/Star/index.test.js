/* eslint-disable no-undef */
import React from 'react';
import renderer from 'react-test-renderer';
import Star from './index';

describe('Star', () => {
  it('renders correctly', () => {
    const tree = renderer.create(<Star />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
