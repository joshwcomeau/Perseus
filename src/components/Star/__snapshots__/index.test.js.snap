exports[`Star renders correctly 1`] = `
<div
  className="starContainer_15kzu1c-o_O-fadeIn_1xfqppp"
  style={
    Object {
      "animationDelay": "519ms",
      "animationDuration": "1550ms",
      "left": undefined,
      "top": undefined,
    }
  }>
  <svg
    className="star_16oth9z-o_O-bright_1fuysop"
    height={NaN}
    style={
      Object {
        "animationDelay": "2977ms",
        "animationDuration": "18.23545110578995s",
      }
    }
    width={NaN}>
    <filter
      height="1000%"
      id="blur-undefined"
      width="1000%"
      x="-100%"
      y="-100%">
      <feGaussianBlur
        in="SourceGraphic"
        stdDeviation={undefined} />
    </filter>
    <circle
      cx={NaN}
      cy={NaN}
      fill="white"
      filter="url(#blur-undefined)"
      r={NaN}
      stroke="#281f60"
      strokeWidth={undefined} />
  </svg>
</div>
`;
